### ISSUE TITLE
Javascript Library and other suggestions

### ISSUE THREAD

**Issue by geoffdavis92** [picturepan2/spectre#24]
First off, love the framework! Really great, clean design, implements many common components flawlessly, yet is incredibly tiny.  (filesize-wise)

Do you have any plans for writing a JS library for interactions with various components, eg `.menu`, `.modal`, and `.toast`? If you haven't or have but have not implemented it yet, I have a few files for said components that could provide a jumping-off point for writing these JS interactions; they're simple, curt, and use vanilla JS.

Other suggestions:

* Include Sass port (this may be able to be done by simply changing all `@variable` instances to `$variable`, but I may be mistaken)
* I resized my window after happily sketching a site in Spectre for a few hours, and I noticed that it's not very responsive relative to the numerical value of column sizes. (ie column widths do not scale consistently with screen size adjustments) [This image](http://prntscr.com/avlxwt) shows that the `.col-8` class, along with apparently all other `.col-x` classes simply default to 50% between 481px and 840px. There may have been a reason for this, but I am not picking up on it.
* Design an adorable/menacing spectre logo for the framework!

I am more than willing to help with any of the above, and will likely be throwing some PRs your way.

**Comment by picturepan2** [comment]
Thank you. I haven't any plan on JS library for `.menu`, `.modal` or `.toast`.
1. I will see if more people need Sass version.
2. Updated this week. Now Spectre provides `.col-xs-`, `col-sm-`, `col-md-`.
3. Sure.

Thank you. 😄

**Comment by geoffdavis92** [comment]
@picturepan2 I cleaned up and updated the javascript that I previously wrote. Still could use some work, but this will give you some idea of what I was shooting for. [Here is the link](https://github.com/geoffdavis92/spectre/tree/javascript/dist/js).

I have 3 files so far, one for menus, modals, and toasts each; technically there could be 4, as I added support for an arrow pseudo element that could switch orientation on click, but I don't know if you wanted that in the project.

* `menu.spectre.js` handles toggling menus, and works best when `.hide` is applied in the menu's html. Currently, only clicking the toggling element will open/close the menu.
* `modal.spectre.js` handles opening modals using a button/link, and closing the modal via either `.btn-clear` button, a link that has `data-target=#idOfModal`/`data-close="modal"`, or clicking the modal overlay.
* `toast.spectre.js` handles closing toasts; this obviously only works on the frontend, and perhaps I will write in a function or method that will allow a callback on closing toasts, so developers can write an API-handling event when a modal closes.

Like I said, these are kind of rough, and only use ES6/vanilla JS. If you like the idea, I can optimize these and write more `.js` files.

**Comment by picturepan2** [comment]
Thank you. I think I will wait and see if other people need JS example codes, maybe just jQuery or Angular examples. Firstly, I need to perfect basic CSS codes.

**Comment by victorwpbastos** [comment]
First, thanks for this amazing project. So simple, clean and powerful! Second, would be great to have a Sass variant (scss).

**Comment by picturepan2** [comment]
@victorwpbastos May I know why you want Sass variant?

**Comment by geoffdavis92** [comment]
@picturepan2 The reason I suggested it originally is because not every one uses LESS, which is fine, and some people would like to compile it during development with custom variable settings.

**Comment by victorwpbastos** [comment]
Sass is just a personal preference. Many CSS frameworks are using Sass instead Less. But I repeat: is just a personal preference.

**Comment by picturepan2** [comment]
Thank you both. Do you know tools to keep Less and Sass coexist and synced?

**Comment by victorwpbastos** [comment]
Maybe maintain the two flavors (Less and Sass) won't be a good idea. If you like Less keep Less. No problem.

**Comment by picturepan2** [comment]
OK. I will keep Less. I will read more about Sass to see if Spectre.css needs.

**Comment by sudheerDev** [comment]
@picturepan2 sass would be an important port! any reason for less implementation?

**Comment by kelkes** [comment]
SASS version would be great.

**Comment by chekun** [comment]
SASS version  +1

**Comment by fabiocaccamo** [comment]
Generally i use Foundation or Bootstrap, and there is a Foundation feature that nobody offers, it's list-block-grid support. In few words if you have a list you can specify how many elements display per each row without breaking the list.

**Comment by IgnisFire** [comment]
No. I dont think SASS version make situation more better. It make things worse! Support two versions is harder. And its make more time to review pull requests and and test everything

**Comment by jakobrosenberg** [comment]
I have seen a lot of projects move from less to sass, but never from sass to less.

> Moved from Less to Sass. Bootstrap now compiles faster than ever thanks to Libsass, and we join an increasingly large community of Sass developers.

To me it's a bit like underscore vs lodash.

**Comment by geoffdavis92** [comment]
@IgnisFire I don't think anyone is suggesting @picturepan2 should support 2 different versions, at least not concurrently. If he were to offer a Sass port and did not want to maintain both, he could release stable Sass versions at major framework versions.

I see the benefit from moving away from LESS to Sass, as @jakobrosenberg pointed out, Sass is faster, and offers (AFAIK) more options for compiling in a variety of languages, including node.js, ruby, python, and more.

On top of this is that a majority of development blogs/articles/developers I'm aware of use Sass to compile their styles into CSS. All these benefits point to Sass being a better pre-compiled language to use for customizing Spectre.css.

**Comment by picturepan2** [comment]
@All, I plan to switch to Sass and release a stable version of Less. Thank you.

**Comment by dak0rn** [comment]
I like the idea of @geoffdavis92 to add JavaScript and I even more like the idea of not using dependencies. However, I looked at the code I'd really recommend not to touch prototypes as it is considered to be (very) bad practice.

Adding some kind of simple bundler (e.g. [Rollup](http://rollupjs.org/)) could be useful to provide some kind of internal structuring without the hassle of using a sophisticated build chain.

**Comment by AdsonCicilioti** [comment]
Sass!! Sounds good!!

**Comment by geoffdavis92** [comment]
@dak0rn Yeah I only added that because I somehow forgot about `Array.forEach`, and was too lazy to look it up.

Sometime soon I'm going to generally clean up the scripts and I'll remove my prototype edit.

**Comment by dak0rn** [comment]
@geoffdavis92 just wanted to mention it.

I hope a possible JavaScript part is not going to target non-ES5.1 environments like the Internet Explorer 8, because that would introduce quite a nasty limitation.

**Comment by groundc** [comment]
I think it's best to leave Javascript out of the picture and for Spectre to remain a focused CSS / Design library while encouraging JS component libraries to be built on top of the framework. Salesforce's design system is a good example of how to go about this. https://www.lightningdesignsystem.com/

**Comment by imanhodjaev** [comment]
Spectre is really simple and I support the idea of keeping Sectre CSS only as well because it is easy to add any existing JS plugin on top of the layout. And supporting SASS would be "just what doctor ordered".

**Comment by geoffdavis92** [comment]
@imanhodjaev That's what I'm feeling is the consensus. 

I'll likely release my library as a "plugin", after cleaning it up and building it out a bit. Not sure how I could brand it, but that's a topic for another time 😄

**Comment by AdsonCicilioti** [comment]
I confess that I just did not begin to use the Spectre because it is not in Sass. It would be great to work with Spetre in Sass.

**Comment by Fantasterei** [comment]
Converting the less-code to sass cost me 5 minutes. You only need to replace the @-sign for variables with the $-sign and add mixins+imports instead of the classes which are used for global usage of styleheets. It can't be easier...

**Comment by larsbo** [comment]
what about the js modules idea?
I think it would be a nice benefit if you can add optional "official" spectre js modules to the page to extend the css functionality of individual elements/components.

**Comment by IlBaffo** [comment]
Please don't make JS an integral part of this framework: I choose it over bootstrap/foundation specifically because it was pure CSS only.

I thinks it is ok to make optional scripts for simple interactions (like a jquery plugin or some vanilla code) and like a reference for implementation but, since Js is so opinionated, writing any kind of code will probably lead to a dozen of variants (react, angular, angular2, kendo,....) with many build systems integrations and feature requests.
I think that this is out of scope for a lightweight library like this and, if forced to reference a Js to use the components, I will just stop updating it and fork it locally.

Just my two cents.

**Comment by geoffdavis92** [comment]
@IlBaffo I agree. 

I'm currently writing a standalone [Javascript library](https://github.com/geoffdavis92/spectre.js) using vanilla js to interact with various spectre.css components. 

I finished writing a preliminary version of handling `.tab.tab-group` last night, will push up in a few hours.

I like spectre for the same reasons you do, but if I can write a simple js library to interact with spectre, I think that'd be helpful to a lot of folks. Not everyone has the time/patience/skill to roll their own. Again, this will be completely optional and unofficial. (I've extended an offer to @picturepan2 to house everything under an org like #50 suggests, but it's entirely up to him. Even still this would just be a completely optional library)

**Comment by aaron-bond** [comment]
@geoffdavis92 I have to +1 the notion of keeping JS out of the equation. Keeping it CSS only means that it can be integrated with any framework you want without much issue. At the minute, I'm demoing dropping it into Angular 2 and it works a treat (tabs handled via bindings)

**Comment by gofabian** [comment]
For me one of the reasons to choose spectre was that it is CSS only. I use bootstrap in another project. The javascript part modifies the HTML elements so that customizing the boostrap elements is not fun.
Please don't add JS support.

**Comment by roemhildtg** [comment]
There's already several different javascript implementations that I know of. (vue, can, etc) Would it be helpful for anyone to have a page in the docs or somewhere called "JS implementations" with links to the projects?

**Comment by aaron-bond** [comment]
I'd contribute to that

**Comment by faststare** [comment]
spectre css logo for attribution link? and where I can find a collaboration partner(who used to be a spectre css hobbyist  for a open source project)?

**Comment by aaron-bond** [comment]
@faststare what's the project?

**Comment by faststare** [comment]
@aaron-bond an accounting web app that have philippine government features(taxation handling and gocc handling).

**Comment by zusamann** [comment]
I think it'll be a great idea to create a `react.spectre` much like how a `react.semantic-ui` exists.
@geoffdavis92 the library you're creating, does that provide any sort of react components, or is it to be used outside react.

**Comment by picturepan2** [comment]
@faststare You need a Spectre.css logo in which sizes?

**Comment by roemhildtg** [comment]
Is there a logo? I'd put one on my project site also.

**Comment by AdsonCicilioti** [comment]
@zusamann React is future.. Or better.. Is Present.. Maybe a good moving idea in this regard.

**Comment by tunecino** [comment]
@zusamann @AdsonCicilioti me, probably like @IlBaffo and many others have chosen this framework because of 2 main great features:

1. It is Lightweight (~10KB gzipped)
2. It has no integral JS

Those 2 facts make it unique, special and different from most of what you can find out there. It is not one more framework trying to do all the work or providing whatever feature you may need like if it was some kind of software a browser need to download on each visit. It is fast because it is simple. Please if you need a Javascript Library for it consider making a fork and adding that to a different repo.

**Comment by tunecino** [comment]
> Again, this will be completely optional and unofficial.

Oh sorry I didn't see that. In that case I 100% agree :)

**Comment by flyboarder** [comment]
Could I also suggest a sticker board, or something like sketchapp.com for designers.

**Comment by picturepan2** [comment]
@flyboarder What is a sticker board?

**Comment by flyboarder** [comment]
Usually a collection of the components for prototyping.

https://material.io/guidelines/resources/sticker-sheets-icons.html

**Comment by picturepan2** [comment]
@geoffdavis92 v0.4.0 is in SASS/SCSS now.
